src/Features/Core/Portable/UnusedReferences/UnusedReferencesService.cs

```diff
@@ -0,0 +1,164 @@
+
+            // Remove all assemblies that are brought into the compilation from those toplevel used references. When
+            // determining transtively used references this will reduce false positives.
+            var remainingReferences = references.Except(toplevelUsedReferences);
+            var remainingUsedAssemblyLookup = usedAssemblyLookup.Except(toplevelUsedAssemblyReferences);
+
+            // Determine which transtive references have compilation assemblies in the set of remaining used assemblies.
+            foreach (var reference in remainingReferences)
+            {
+                var allCompilationAssemblies = reference.GetAllCompilationAssemblies();
+                if (allCompilationAssemblies.IsEmpty)
+                {
+                    // We will consider References that do not contribute any assemblies to the compilation,
+                    // such as Analyzer packages, as used.
+                    continue;
+                }
+
+                if (!allCompilationAssemblies.Any(remainingUsedAssemblyLookup.Contains))
+                {
+                    // None of the assemblies brought into this compilation are in the remaining
+                    // used assemblies list, so we will consider the reference unused.
+                    unusedReferences.Add(reference);
+                    continue;
+                }
+
+                remainingUsedAssemblyLookup = remainingUsedAssemblyLookup.Except(allCompilationAssemblies);
+            }
+
+            return remainingUsedAssemblyLookup;
+        }
+
+        public async Task<Project> UpdateReferencesAsync(
+            Project project,
+            ImmutableArray<ReferenceUpdate> referenceUpdates,
+            CancellationToken cancellationToken)
+        {
+            var referenceCleanupService = project.Solution.Workspace.Services.GetRequiredService<IReferenceCleanupService>();
+
+            await ApplyReferenceUpdatesAsync(referenceCleanupService, project.FilePath!, referenceUpdates, cancellationToken).ConfigureAwait(false);
+
+            return project.Solution.Workspace.CurrentSolution.GetRequiredProject(project.Id);
+        }
+
+        internal static async Task ApplyReferenceUpdatesAsync(
+            IReferenceCleanupService referenceCleanupService,
+            string projectFilePath,
+            ImmutableArray<ReferenceUpdate> referenceUpdates,
+            CancellationToken cancellationToken)
+        {
+            foreach (var referenceUpdate in referenceUpdates)
+            {
+                // If the update action would not change the reference, then
+                // continue to the next update.
+                if (referenceUpdate.Action == UpdateAction.TreatAsUnused &&
+                    !referenceUpdate.ReferenceInfo.TreatAsUsed)
+                {
+                    continue;
+                }
+                else if (referenceUpdate.Action == UpdateAction.TreatAsUsed &&
```

Comment:
Did you mean this check to be `!=` instead of `==`?

Reply:
Not in this case.  We are looking, when the action is TreatAsUsed,  to see if our reference is already marked as TreatAsUsed.